api/src/test/java/org/pmiops/workbench/billing/FreeTierBillingServiceTest.java

```diff
@@ -55,221 +59,617 @@
       return workbenchConfig;
     }
   }
 
   @Before
-  public void setUp() throws Exception {
+  public void setUp() {
     workbenchConfig = WorkbenchConfig.createEmptyConfig();
+  }
+
+  @Test
+  public void checkFreeTierBillingUsage_exceedsDollarThresholds() {
+    final double limit = 100.0;
+    final double costUnderThreshold = 49.5;
+
+    double threshold = 0.5;
+    double costOverThreshold = 50.5;
+    double remaining = limit - costOverThreshold;
+
+    workbenchConfig.billing.defaultFreeCreditsDollarLimit = limit;
+
+    final DbUser user = createUser(SINGLE_WORKSPACE_TEST_USER);
+    createWorkspace(user, SINGLE_WORKSPACE_TEST_PROJECT);
+
+    // check that we have not alerted before the threshold
+
+    doReturn(mockBQTableSingleResult(costUnderThreshold)).when(bigQueryService).executeQuery(any());
+    freeTierBillingService.checkFreeTierBillingUsage();
+    verifyZeroInteractions(notificationService);
+
+    // check that we alert for the 50% threshold
+
+    doReturn(mockBQTableSingleResult(costOverThreshold)).when(bigQueryService).executeQuery(any());
+    freeTierBillingService.checkFreeTierBillingUsage();
+    verify(notificationService)
+        .alertUserFreeTierDollarThreshold(
+            eq(user), eq(threshold), eq(costOverThreshold), eq(remaining));
+
+    // check that we do not alert twice for the 50% threshold
+
+    doReturn(mockBQTableSingleResult(costOverThreshold)).when(bigQueryService).executeQuery(any());
+    freeTierBillingService.checkFreeTierBillingUsage();
+    verifyZeroInteractions(notificationService);
+
+    // check that we alert for the 75% threshold
+
+    threshold = 0.75;
+    costOverThreshold = 75.3;
+    remaining = limit - costOverThreshold;
+
+    doReturn(mockBQTableSingleResult(costOverThreshold)).when(bigQueryService).executeQuery(any());
+    freeTierBillingService.checkFreeTierBillingUsage();
+    verify(notificationService)
+        .alertUserFreeTierDollarThreshold(
+            eq(user), eq(threshold), eq(costOverThreshold), eq(remaining));
+
+    // check that we do not alert twice for the 75% threshold
+
+    doReturn(mockBQTableSingleResult(costOverThreshold)).when(bigQueryService).executeQuery(any());
+    freeTierBillingService.checkFreeTierBillingUsage();
+    verifyZeroInteractions(notificationService);
+
+    // check that we alert for expiration when we hit 100%
+
+    final double costToTriggerExpiration = 100.01;
+
+    doReturn(mockBQTableSingleResult(costToTriggerExpiration))
+        .when(bigQueryService)
+        .executeQuery(any());
+    freeTierBillingService.checkFreeTierBillingUsage();
+    verify(notificationService).alertUserFreeTierExpiration(eq(user));
+
+    // check that we do not alert twice for 100%
+
+    doReturn(mockBQTableSingleResult(costToTriggerExpiration))
+        .when(bigQueryService)
+        .executeQuery(any());
+    freeTierBillingService.checkFreeTierBillingUsage();
+    verifyZeroInteractions(notificationService);
```

Comment:
great test! Easy to understand and I really like how we're testing it from the BQ input to the Alerting output interfaces.